autopilot/hostinfo.go

```diff
@@ -65,66 +65,72 @@ func (c *contractor) HostInfo(ctx context.Context, hostKey types.PublicKey) (api
-	}
-	if cfg.Contracts.Amount == 0 {
-		return nil, fmt.Errorf("can not score hosts because contracts amount is zero")
-	}
-	if cfg.Contracts.Period == 0 {
-		return nil, fmt.Errorf("can not score hosts because contract period is zero")
-	}
+func (c *contractor) HostInfos(ctx context.Context, filterMode, usabilityMode, addressContains string, keyIn []types.PublicKey, offset, limit int) ([]api.HostHandlerGET, error) {
+	c.mu.Lock()
+	hostInfo := c.cachedHostInfo
+	c.mu.Unlock()
 
-	hosts, err := c.ap.bus.SearchHosts(ctx, filterMode, addressContains, keyIn, offset, limit)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch requested hosts from bus: %w", err)
-	}
-	gs, err := c.ap.bus.GougingSettings(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch gouging settings from bus: %w", err)
-	}
-	rs, err := c.ap.bus.RedundancySettings(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch redundancy settings from bus: %w", err)
-	}
-	cs, err := c.ap.bus.ConsensusState(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch consensus state from bus: %w", err)
-	}
-	fee, err := c.ap.bus.RecommendedFee(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch recommended fee from bus: %w", err)
+	// declare helper to decide whether to keep a host.
+	if usabilityMode != api.UsabilityFilterModeUsable && usabilityMode != api.UsabilityFilterModeUnusable && usabilityMode != api.UsabilityFilterModeAll {
+		return nil, fmt.Errorf("invalid usability mode: %v, options are usable and unusable", usabilityMode)
+	}
+	keep := func(usable bool) bool {
+		switch usabilityMode {
+		case api.UsabilityFilterModeUsable:
+			return usable // keep usable
+		case api.UsabilityFilterModeUnusable:
+			return !usable // keep unusable
+		case api.UsabilityFilterModeAll:
+			return true // keep all
+		default:
+			panic("unreachable")
+		}
 	}
 
-	f := newIPFilter(c.logger)
-	gc := worker.NewGougingChecker(gs, rs, cs, fee, cfg.Contracts.Period, cfg.Contracts.RenewWindow)
+	var hostInfos []api.HostHandlerGET
+	wanted := limit
+	for {
+		// fetch up to 'limit' hosts.
+		hosts, err := c.ap.bus.SearchHosts(ctx, filterMode, addressContains, keyIn, offset, limit)
+		if err != nil {
+			return nil, err
+		}
+		offset += len(hosts)
 
-	c.mu.Lock()
-	storedData := make(map[types.PublicKey]uint64)
-	for _, host := range hosts {
-		storedData[host.PublicKey] = c.cachedDataStored[host.PublicKey]
-	}
-	minScore := c.cachedMinScore
-	c.mu.Unlock()
+		// if there are no more hosts, we're done.
+		if len(hosts) == 0 {
+			return hostInfos, nil // no more hosts
+		}
 
-	var hostInfos []api.HostHandlerGET
-	for _, host := range hosts {
-		// ignore the pricetable's HostBlockHeight by setting it to our own blockheight
-		host.PriceTable.HostBlockHeight = cs.BlockHeight
+		// decide how many of the returned hosts to keep.
+		var keptHosts int
+		for _, host := range hosts {
+			hi, cached := hostInfo[host.PublicKey]
+			if !cached {
+				continue
```

Comment:
I'm wondering whether we should return an empty `hostInfo` here perhaps, like return the host key without usability info? Not sure, it's just weird to just skip hosts which aren't in the cache. I'm pretty sure this will lead to all sorts of questions like "hey my host is missing in the infos" or things of that nature.

Reply:
I considered it but at the same time it will only take 10 minutes at most for the host to end up in the cache after first appearing on chain. Before appearing in the cache the host can never be usable.